src/modules/job-manager/plugins/jobspec-default.c

```diff
@@ -0,0 +1,414 @@
+                      const char *topic,
+                      flux_plugin_arg_t *args,
+                      void *arg)
+{
+    flux_t *h = flux_jobtap_get_flux (p);
+    flux_job_state_t state;
+    json_t *conf = NULL;
+    json_t *jobspec = NULL;
+    json_t *defaults = NULL;
+    json_t *update = NULL;
+    flux_error_t error;
+
+    if (flux_plugin_arg_unpack (args,
+                                FLUX_PLUGIN_ARG_IN,
+                                "{s:i s:{s:{s?o}}}",
+                                "state", &state,
+                                "jobspec",
+                                  "attributes",
+                                    "system", &jobspec) < 0) {
+        flux_jobtap_reject_job (p,
+                                args,
+                                "error unpacking job.create arguments: %s",
+                                flux_plugin_arg_strerror (args));
+        goto error;
+    }
+
+    /* If state is not NEW, this is a job manager/flux restart and any
+     * defaults will have already been replayed from the KVS.
+     */
+    if (state != FLUX_JOB_STATE_NEW)
+        goto done;
+
+    (void)flux_conf_unpack (flux_get_conf (h), NULL, "o", &conf);
+
+    /* Construct defaults dict by overlaying general and queue specific
+     * configured defaults.
+     */
+    if (get_policy_defaults (conf, jobspec, &defaults, &error) < 0) {
+        flux_jobtap_reject_job (p, args, "%s", error.text);
+        flux_log (h, LOG_ERR, "jobspec-default: %s", error.text);
+        goto error;
+    }
+
+    /* If no defaults are configured, there is nothing to do.
+     */
+    if (!defaults)
+        goto done;
+
+    /* Walk the configured default system attributes.
+     * If an attribute appears in jobspec, leave it alone.
+     * If it is missing from jobspec, add it to the 'update' object.
+     */
+    if (!(update = generate_update (defaults, jobspec))) {
+        flux_jobtap_reject_job (p,
+                                args,
+                                "error creating jobspec-update context: %s",
+                                 strerror (errno));
+        goto error;
+    }
+
+    /* Post jobspec-update event, if updates were generated.
+     */
+    if (json_object_size (update) > 0) {
+        if (flux_jobtap_event_post_pack (p,
+                                         FLUX_JOBTAP_CURRENT_JOB,
+                                         "jobspec-update",
+                                         "O",
+                                         update) < 0) {
+            flux_jobtap_reject_job (p,
+                                    args,
+                                    "failed to post jobspec-update: %s",
+                                    strerror (errno));
+            goto error;
+        }
+    }
+done:
+    json_decref (defaults);
+    json_decref (update);
+    return 0;
+error:
+    json_decref (defaults);
+    json_decref (update);
+    return -1;
```

Comment:
Here was my one style comment: This code duplication stuck out to me. Would it be cleaner to initialize an `rc` variable at the top and then remove the duplication via something like

```c
int rc = -1;
```

```c
    rc = 0;
error:    
    json_decref (defaults);
    json_decref (update);
    return rc;
```

Or do people not like seeing a normal return after an `error` label? (In that case maybe rename the label `out`.)

Reply:
No, that's a fair point. To be honest I've been avoiding the "rc" idiom in all but the simplest cases after I spent some time unwinding the error handling in some flux-sched functions a long time ago that "re-used" rc to hold internal function call results and fell through to the end without gotos.  When the control flow is clear, it is not a problem so I probably overreacted.

Reply:
That is a good point, and I've probably been guilty of that `rc` misuse in trying to avoid the duplicate `free` blocks (hopefully not too often though!). We just need a version of C with the [`defer`](https://open-std.org/jtc1/sc22/wg14/www/docs/n2895.htm) keyword.